### DISCUSSION TITLE
[API] Form hooks

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#36523] [RFCs]
## Related Issues

* #35036
* #36003

### Form Level

Which appears for the form level control

#### ✅ Form.useForm

Create Form instance which store the data info

#### ✅ Form.useWatch

Listen for specific name path value change state

#### 🉑 Form.useMeta

Listen for Field meta change like errors, touched, dirty, validating

#### 🚫 Form.useError

Can use `Form.useMeta` instead

#### 🚫 Form.useTouched

Can use `Form.useMeta` instead

#### 🚫 Form.useDirty

Can use `Form.useMeta` instead

#### 🚫 Form.useValidating

Can use `Form.useMeta` instead

-----

### FormItem Level

Which only for consuming Form Item info. This hooks will only work under `<Form.Item />`!

#### 🉑 FormItem.useStatus

Get parent Form.Item validate status which will get the status & icon

**Comment by zombieJ** [comment]
`useErrors` `useTouched` 之类的其实也可以有，就是看要不要拆的这么细。粗一点的 `useMeta` 可以把边角料都关注到，细的话就会控制更精准一些。

**Comment by MadCcc** [comment]
> FormItem.useStatus

实际用起来就变成 `Form.Item.useStatus()` 了，不过解构之后倒也差别不大

**Comment by zombieJ** [comment]
嗯，我想这边和 Form 的 hooks 做一个区分。放一起很容易让人误解是监听级别的 hooks。FormItem 的 hooks 应该是只满足少部分自定义需求的人才用到

**Comment by afc163** [comment]
想到的可以都先写出来，不一定都要做。

**Comment by afc163** [comment]
useErrors 和 useStatus 的区别是？

**Comment by zombieJ** [comment]
> useErrors 和 useStatus 的区别是？

`Form.useError` 是字段的 error 变化触发 rerender，`FormItem.useStatus` 是直接消费来自 FormItem 的 Context 里的 status 信息用于渲染展示（是用于自定义组件消费 FormItem 内部数据）：
<img width="365" alt="截屏2022-07-13 上午10 50 40" src="https://user-images.githubusercontent.com/5378891/178640420-666646fe-727a-41dd-9427-e4f188983305.png">

**Comment by crazyair** [comment]
消费的话，我一般都是直接用 useContext(xxx)，这个还需要单独封装 useStatus hooks 吗？

**Comment by MadCcc** [comment]
> 消费的话，我一般都是直接用 useContext(xxx)，这个还需要单独封装 useStatus hooks 吗？

防止 context 实例不同，目前没有正常透出 context

**Comment by crazyair** [comment]
`const { status } = useContext(Form.Item.Context)`

**Comment by zombieJ** [comment]
> 消费的话，我一般都是直接用 useContext(xxx)，这个还需要单独封装 useStatus hooks 吗？

开头我也觉得 Context 好，后来想想 Context 容易在维护的时候不小心加了什么额外东西透出去导致 API 泛滥。干脆封装掉，这样不会污染 API

**Comment by crazyair** [comment]
> > 消费的话，我一般都是直接用 useContext(xxx)，这个还需要单独封装 useStatus hooks 吗？
> 
> 开头我也觉得 Context 好，后来想想 Context 容易在维护的时候不小心加了什么额外东西透出去导致 API 泛滥。干脆封装掉，这样不会污染 API

可以加单元测试，确定导出的参数。这样防止被意外增加。

**Comment by zombieJ** [comment]
> 可以加单元测试，确定导出的参数。这样防止被意外增加。

https://github.com/ant-design/ant-design/pull/36486#issuecomment-1181827653  最早我也是这么想的，后来觉得既然入口就可以约束掉。就不用等到测试阶段了。
